File: crates/up_down/contracts/common/src/pool_data.rs
//! PoolCell data layout and manual little-endian encoding.
//!
//! `pool_id` lives in the type-script args (the typeID), not here. The UP/DOWN
//! share-token and treasury identities are config, not build-time constants. The
//! layout is led by `variant`, which decides whether xUDT-only fields are present:
//!
//! - CKB variant  (`variant == 0`): `POOL_LEN_CKB  = 173`
//! - xUDT variant (`variant == 1`): `POOL_LEN_XUDT = 237`
//!
//! See `docs/pool_type-spec.md` for the full offset tables.

use crate::constants::{VARIANT_CKB, VARIANT_XUDT};
use alloc::vec::Vec;

pub const POOL_LEN_CKB: usize = 173;
pub const POOL_LEN_XUDT: usize = 237;

/// The contiguous oracle-identity block that follows the script-identity config:
/// feed_id(32) oracle_commit(32) = 64. `oracle_commit` is a single hash binding
/// the oracle type code hash + the trust root (guardian-set type hash + Pyth
/// emitter chain/address) — see `oracle_read::oracle_commit`. `feed_id` stays in
/// the clear because it's the oracle type's args (used to locate the cell).
const ORACLE_ID_LEN: usize = 64;

/// Bytes after the oracle-identity block: start_time(8) close_time(8)
/// up_total(16) down_total(16) start_price(8) settle_price(8) used_pt(8)
/// rake_bps(2) status(1) winner(1) = 76.
const TAIL_LEN: usize = 76;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PoolData {
    pub variant: u8,
    /// Some iff `variant == VARIANT_XUDT`.
    pub asset_type_hash: Option<[u8; 32]>,
    /// Code hash of this pool's UP/DOWN share xUDT script.
    pub share_xudt_code_hash: [u8; 32],
    /// Code hash of this pool's treasury lock, present iff `variant == VARIANT_XUDT`.
    pub treasury_lock_code_hash: Option<[u8; 32]>,
    /// Pyth feed id — the oracle type script's args. Pins which feed *and*,
    /// together with `oracle_code_hash`, exactly which oracle type is trusted.
    pub feed_id: [u8; 32],
    /// Commitment to the trusted oracle's identity: a single hash binding the
    /// oracle type code hash + trust root (guardian-set type hash, Pyth emitter
    /// chain/address). `find_oracle` recomputes it from the dep cell and matches.
    /// See `oracle_read::oracle_commit`.
    pub oracle_commit: [u8; 32],
    pub start_time: u64,
    pub close_time: u64,
    pub up_total: u128,
    pub down_total: u128,
    pub start_price: i64,
    pub settle_price: i64,
    /// `publish_time` of the oracle tick backing the price of the *current phase*.
    /// In LOCKED it tracks the **start** tick (in `(start,close)`); RESOLVE flips it
    /// to the **settle** tick (in `(close,void)`). Within a phase a CORRECT may only
    /// replace it with a strictly *earlier* tick (monotone-down convergence toward the
    /// first tick after the boundary). State, not config. Zero until activation.
    pub used_pt: u64,
    pub rake_bps: u16,
    pub status: u8,
    pub winner: u8,
}

impl PoolData {
    /// Decode PoolCell data. Peeks `variant`, then branches on length.
    pub fn from_bytes(d: &[u8]) -> Option<Self> {
        let variant = *d.first()?;
        let (asset_type_hash, treasury_lock_code_hash, r0) = match variant {
            VARIANT_CKB => (None, None, 33usize),
            VARIANT_XUDT => {
                let mut h = [0u8; 32];
                h.copy_from_slice(d.get(1..33)?);
                let mut t = [0u8; 32];
                t.copy_from_slice(d.get(65..97)?);
                (Some(h), Some(t), 97usize)
            }
            _ => return None,
        };

        if d.len() != r0 + ORACLE_ID_LEN + TAIL_LEN {
            return None;
        }

        let share_off = match variant {
            VARIANT_CKB => 1usize,
            VARIANT_XUDT => 33usize,
            _ => return None,
        };
        let mut share_xudt_code_hash = [0u8; 32];
        share_xudt_code_hash.copy_from_slice(d.get(share_off..share_off + 32)?);

        let mut feed_id = [0u8; 32];
        feed_id.copy_from_slice(&d[r0..r0 + 32]);
        let mut oracle_commit = [0u8; 32];
        oracle_commit.copy_from_slice(&d[r0 + 32..r0 + 64]);
        let r = r0 + ORACLE_ID_LEN;

        Some(Self {
            variant,
            asset_type_hash,
            share_xudt_code_hash,
            treasury_lock_code_hash,
            feed_id,
            oracle_commit,
            start_time: u64::from_le_bytes(d[r..r + 8].try_into().ok()?),
            close_time: u64::from_le_bytes(d[r + 8..r + 16].try_into().ok()?),
            up_total: u128::from_le_bytes(d[r + 16..r + 32].try_into().ok()?),
            down_total: u128::from_le_bytes(d[r + 32..r + 48].try_into().ok()?),
            start_price: i64::from_le_bytes(d[r + 48..r + 56].try_into().ok()?),
            settle_price: i64::from_le_bytes(d[r + 56..r + 64].try_into().ok()?),
            used_pt: u64::from_le_bytes(d[r + 64..r + 72].try_into().ok()?),
            rake_bps: u16::from_le_bytes(d[r + 72..r + 74].try_into().ok()?),
            status: d[r + 74],
            winner: d[r + 75],
        })
    }

    /// Encode back into the exact byte layout.
    pub fn to_bytes(&self) -> Vec<u8> {
        let cap = if self.variant == VARIANT_XUDT {
            POOL_LEN_XUDT
        } else {
            POOL_LEN_CKB
        };
        let mut out = Vec::with_capacity(cap);
        out.push(self.variant);
        if let Some(h) = &self.asset_type_hash {
            out.extend_from_slice(h);
        }
        out.extend_from_slice(&self.share_xudt_code_hash);
        if let Some(h) = &self.treasury_lock_code_hash {
            out.extend_from_slice(h);
        }
        out.extend_from_slice(&self.feed_id);
        out.extend_from_slice(&self.oracle_commit);
        out.extend_from_slice(&self.start_time.to_le_bytes());
        out.extend_from_slice(&self.close_time.to_le_bytes());
        out.extend_from_slice(&self.up_total.to_le_bytes());
        out.extend_from_slice(&self.down_total.to_le_bytes());
        out.extend_from_slice(&self.start_price.to_le_bytes());
        out.extend_from_slice(&self.settle_price.to_le_bytes());
        out.extend_from_slice(&self.used_pt.to_le_bytes());
        out.extend_from_slice(&self.rake_bps.to_le_bytes());
        out.push(self.status);
        out.push(self.winner);
        out
    }

    /// Config fields that must never change after creation.
    pub fn config_unchanged(&self, o: &Self) -> bool {
        self.variant == o.variant
            && self.asset_type_hash == o.asset_type_hash
            && self.share_xudt_code_hash == o.share_xudt_code_hash
            && self.treasury_lock_code_hash == o.treasury_lock_code_hash
            && self.feed_id == o.feed_id
            && self.oracle_commit == o.oracle_commit
            && self.start_time == o.start_time
            && self.close_time == o.close_time
            && self.rake_bps == o.rake_bps
    }

    /// Pool duration in seconds.
    pub fn duration(&self) -> u64 {
        self.close_time.saturating_sub(self.start_time)
    }
}

File: packages/game-sdk/src/codec/poolData.ts
//! PoolData encode/decode — a byte-for-byte mirror of
//! `crates/up_down/contracts/common/src/pool_data.rs`. The layout is led by
//! `variant`, which decides whether xUDT-only fields are present:
//!
//!   variant(1) [asset_type_hash(32) if xUDT] share_xudt_code_hash(32)
//!   [treasury_lock_code_hash(32) if xUDT] feed_id(32) oracle_commit(32)
//!   start_time(8) close_time(8) up_total(16) down_total(16)
//!   start_price(8, i64) settle_price(8, i64) used_pt(8) rake_bps(2) status(1) winner(1)
//!
//! CKB = 173 bytes, xUDT = 237 bytes.

import {
  POOL_LEN_CKB,
  POOL_LEN_XUDT,
  RAKE_BPS_MAX,
  VARIANT_CKB,
  VARIANT_XUDT,
} from "../constants.js";
import {
  bytesToHex,
  getU128LE,
  hexToFixed,
  setU128LE,
  type Hex,
} from "../internal/bytes.js";
import type { PoolData } from "../types.js";

const U64_MAX = (1n << 64n) - 1n;

function requireU64(name: string, v: bigint): bigint {
  if (v < 0n || v > U64_MAX) throw new Error(`${name} out of u64 range: ${v}`);
  return v;
}

function requireU8(name: string, v: number): number {
  if (!Number.isInteger(v) || v < 0 || v > 0xff) throw new Error(`${name} out of u8 range: ${v}`);
  return v;
}

export function encodePoolData(pd: PoolData): Uint8Array {
  const isXudt = pd.variant === VARIANT_XUDT;
  if (pd.variant !== VARIANT_CKB && !isXudt) {
    throw new Error(`unknown variant: ${pd.variant}`);
  }
  if (isXudt && !pd.assetTypeHash) {
    throw new Error("xUDT variant requires assetTypeHash");
  }
  if (!isXudt && pd.assetTypeHash) {
    throw new Error("CKB variant must not carry assetTypeHash");
  }
  if (!pd.shareXudtCodeHash) {
    throw new Error("shareXudtCodeHash is required");
  }
  if (isXudt && !pd.treasuryLockCodeHash) {
    throw new Error("xUDT variant requires treasuryLockCodeHash");
  }
  if (!isXudt && pd.treasuryLockCodeHash) {
    throw new Error("CKB variant must not carry treasuryLockCodeHash");
  }
  if (!Number.isInteger(pd.rakeBps) || pd.rakeBps < 0 || pd.rakeBps > RAKE_BPS_MAX) {
    throw new Error(`rakeBps out of range [0, ${RAKE_BPS_MAX}]: ${pd.rakeBps}`);
  }

  const len = isXudt ? POOL_LEN_XUDT : POOL_LEN_CKB;
  const out = new Uint8Array(len);
  const dv = new DataView(out.buffer);

  let o = 0;
  out[o] = requireU8("variant", pd.variant);
  o += 1;
  if (isXudt) {
    out.set(hexToFixed(pd.assetTypeHash as string, 32, "assetTypeHash"), o);
    o += 32;
  }
  out.set(hexToFixed(pd.shareXudtCodeHash, 32, "shareXudtCodeHash"), o);
  o += 32;
  if (isXudt) {
    out.set(hexToFixed(pd.treasuryLockCodeHash as string, 32, "treasuryLockCodeHash"), o);
    o += 32;
  }
  out.set(hexToFixed(pd.feedId, 32, "feedId"), o);
  o += 32;
  out.set(hexToFixed(pd.oracleCommit, 32, "oracleCommit"), o);
  o += 32;
  dv.setBigUint64(o, requireU64("startTime", pd.startTime), true);
  o += 8;
  dv.setBigUint64(o, requireU64("closeTime", pd.closeTime), true);
  o += 8;
  setU128LE(dv, o, pd.upTotal);
  o += 16;
  setU128LE(dv, o, pd.downTotal);
  o += 16;
  dv.setBigInt64(o, BigInt.asIntN(64, pd.startPrice), true);
  o += 8;
  dv.setBigInt64(o, BigInt.asIntN(64, pd.settlePrice), true);
  o += 8;
  dv.setBigUint64(o, requireU64("usedPt", pd.usedPt), true);
  o += 8;
  dv.setUint16(o, pd.rakeBps, true);
  o += 2;
  out[o] = requireU8("status", pd.status);
  o += 1;
  out[o] = requireU8("winner", pd.winner);
  o += 1;

  return out;
}

export function encodePoolDataHex(pd: PoolData): Hex {
  return bytesToHex(encodePoolData(pd));
}

/** Decode PoolCell data. Returns `null` on unknown variant or wrong length. */
export function decodePoolData(input: Uint8Array | string): PoolData | null {
  const d = typeof input === "string" ? hexToFixedLoose(input) : input;
  if (d.length === 0) return null;
  const variant = d[0];

  let r0: number;
  let assetTypeHash: Hex | undefined;
  let treasuryLockCodeHash: Hex | undefined;
  let shareOffset: number;
  if (variant === VARIANT_CKB) {
    shareOffset = 1;
    r0 = 33;
  } else if (variant === VARIANT_XUDT) {
    if (d.length < 97) return null;
    assetTypeHash = bytesToHex(d.subarray(1, 33));
    shareOffset = 33;
    treasuryLockCodeHash = bytesToHex(d.subarray(65, 97));
    r0 = 97;
  } else {
    return null;
  }

  const expected = r0 + 64 + 76; // oracle-id block (64) + tail (76)
  if (d.length !== expected) return null;

  const dv = new DataView(d.buffer, d.byteOffset, d.byteLength);
  const shareXudtCodeHash = bytesToHex(d.subarray(shareOffset, shareOffset + 32));
  const feedId = bytesToHex(d.subarray(r0, r0 + 32));
  const oracleCommit = bytesToHex(d.subarray(r0 + 32, r0 + 64));
  let r = r0 + 64;

  const startTime = dv.getBigUint64(r, true);
  const closeTime = dv.getBigUint64(r + 8, true);
  const upTotal = getU128LE(dv, r + 16);
  const downTotal = getU128LE(dv, r + 32);
  const startPrice = dv.getBigInt64(r + 48, true);
  const settlePrice = dv.getBigInt64(r + 56, true);
  const usedPt = dv.getBigUint64(r + 64, true);
  const rakeBps = dv.getUint16(r + 72, true);
  const status = d[r + 74];
  const winner = d[r + 75];

  return {
    variant,
    assetTypeHash,
    shareXudtCodeHash,
    treasuryLockCodeHash,
    feedId,
    oracleCommit,
    startTime,
    closeTime,
    upTotal,
    downTotal,
    startPrice,
    settlePrice,
    usedPt,
    rakeBps,
    status,
    winner,
  };
}

function hexToFixedLoose(hex: string): Uint8Array {
  const s = hex.startsWith("0x") ? hex.slice(2) : hex;
  if (s.length % 2 !== 0) throw new Error(`odd-length hex: ${hex}`);
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(s.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}
